src/probnum/diffeq/odefiltsmooth/initialize.py

```diff
@@ -0,0 +1,318 @@
+        if order > 1:
+            initmean[2 :: (order + 1)] = df(t0, z0) @ f(t0, z0)
+            initcov_diag[2 :: (order + 1)] = SMALL_VALUE
+
+    initcov = np.diag(initcov_diag)
+    initcov_cholesky = np.diag(np.sqrt(initcov_diag))
+    initrv = pnrv.Normal(initmean, initcov, cov_cholesky=initcov_cholesky)
+    kalman = pnfs.Kalman(prior, measmod, initrv)
+
+    out = kalman.filtsmooth(ys, ts)
+
+    estimated_initrv = out.state_rvs[0]
+    mean = estimated_initrv.mean
+    std = estimated_initrv.std
+    return mean, std
+
+
+def compute_all_derivatives_via_taylormode(f, z0, t0, order):
+    """Compute derivatives of the initial conditions of an IVP with Taylor-mode
+    automatic differentiation.
+
+    This requires JAX. For an explanation of what happens ``under the hood``, see [1]_.
+
+    References
+    ----------
+    .. [1] Krämer, N. and Hennig, P., Stable implementation of probabilistic ODE solvers,
+       *arXiv:2012.10106*, 2020.
+
+
+    The implementation is inspired by the implementation in
+    https://github.com/jacobjinkelly/easy-neural-ode/blob/master/latent_ode.py
+
+    Parameters
+    ----------
+    f
+        ODE vector field.
+    z0
+        Initial value.
+    t0
+        Initial time point.
+    order
+        Number of derivatives to compute. Corresponds to the order of the prior (for instance the number of times an integrated Brownian motion process is integrated).
+
+    Returns
+    -------
+    array
+        Estimate of the full stack of derivatives
+    array
+        Array full of zeros, as a dummy error estimate of the estimation.
+    """
+
+    try:
+        import jax.numpy as jnp
+        from jax.config import config
+        from jax.experimental.jet import jet
+
+        config.update("jax_enable_x64", True)
+    except ImportError as err:
+        raise ImportError(
+            "Cannot perform Taylor-mode initialisation without optional "
+            "dependencies jax and jaxlib. Try installing them via `pip install jax jaxlib`."
+        ) from err
+
+    def total_derivative(z_t):
+        """Total derivative."""
+        z, t = jnp.reshape(z_t[:-1], z_shape), z_t[-1]
+        dz = jnp.ravel(f(t, z))
+        dt = jnp.array([1.0])
+        dz_t = jnp.concatenate((dz, dt))
+        return dz_t
+
+    z_shape = z0.shape
+    z_t = jnp.concatenate((jnp.ravel(z0), jnp.array([t0])))
+
+    derivs = []
+
+    derivs.extend(z0)
+    if order == 0:
+        all_derivs = pnss.convert_derivwise_to_coordwise(
+            jnp.array(derivs), ordint=0, spatialdim=len(z0)
+        )
+
+        return all_derivs, jnp.zeros(len(derivs))
+    (y0, [*yns]) = jet(total_derivative, (z_t,), ((jnp.ones_like(z_t),),))
+    derivs.extend(y0[:-1])
+    if order == 1:
+        all_derivs = pnss.convert_derivwise_to_coordwise(
+            jnp.array(derivs), ordint=1, spatialdim=len(z0)
+        )
+
+        return all_derivs, jnp.zeros(len(derivs))
+
+    order = order - 2
+    for _ in range(order + 1):
```

Comment:
do you mean having a case distinction for ``order==0`` and another one for ``order > 0``?

Reply:
Hmmm in principle, I guess so. It is not suuuper easy though, because all the calls to `jet` are different. How about we postpone cleaning up this function?